### PR TITLE
4803 optimize latex show urls

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Refactor

### Purpose
- see #4803 

### Result

```
# before (HEAD of 1.7)
.tox/py36/bin/sphinx-build -b latex doc build/sphinx/latex  69.88s user 1.18s system 92% cpu 1:16.55 total
.tox/py36/bin/sphinx-build -b latex doc build/sphinx/latex  64.74s user 0.63s system 97% cpu 1:07.04 total
.tox/py36/bin/sphinx-build -b latex doc build/sphinx/latex  64.62s user 0.54s system 98% cpu 1:06.43 total

# after (this PR)
.tox/py36/bin/sphinx-build -b latex doc build/sphinx/latex  18.01s user 0.58s system 89% cpu 20.751 total
.tox/py36/bin/sphinx-build -b latex doc build/sphinx/latex  17.95s user 0.51s system 89% cpu 20.562 total
.tox/py36/bin/sphinx-build -b latex doc build/sphinx/latex  16.42s user 0.42s system 94% cpu 17.761 total
```

### CODE REVIEW

**Comment by tk0miya** [comment]
Note:
In this PR, `sphinx/builders/latex.py` is renamed to `sphinx/builders/latex/__init__.py`.
I think this is not breaking change.

File: sphinx/builders/latex/transforms.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        footnote_ref = nodes.footnote_reference('[#]_', label, auto=1,
+                                                refid=footnote['ids'][0])
+        self.document.note_autofootnote_ref(footnote_ref)
+        footnote.add_backref(footnote_ref['ids'][0])
+
+        return [footnote, footnote_ref]
+
+    def renumber_footnotes(self):
+        # type: () -> None
+        collector = FootnoteCollector(self.document)
```
Comment (tk0miya) [inline sphinx/builders/latex/transforms.py:88]: I introduced `FootnoteCollector` to make renumbering faster!

**Comment by jfbu** [comment]
AMAZING speed improvement!

**Comment by tk0miya** [comment]
Thank you for reviewing! :-)

**Comment by jfbu** [comment]
Even on smaller project than Sphinx own docs I get very significant speed improvement for ``make latex``, for example going down from 10 seconds to 5 seconds for a project using ``latex_show_urls     = 'footnote'`` and ending up with about 120 such footnotes (in a 180 pages PDF after the PDF build)

This is very very satisfactory improvement!

**Comment by tk0miya** [comment]
:-)

**Pull request merged**